HSBM.py

```diff
@@ -0,0 +1,391 @@
+import numpy as np
+import pandas as pd
+import warnings
+
+from anytree import NodeMixin, LevelOrderGroupIter, LevelOrderIter
+# from scipy.cluster.hierarchy import dendrogram, linkage
+# from sklearn.cluster import AgglomerativeClustering
+
+from graspy.embed import AdjacencySpectralEmbed, LaplacianSpectralEmbed, select_dimension
+from graspy.cluster import KMeansCluster, AutoGMMCluster
+# from graspy.cluster.base import BaseCluster
+# from graspy.models import SBMEstimator
+from graspy.models.sbm import _calculate_block_p, _block_to_full, _get_block_indices
+from graspy.models.base import BaseGraphEstimator
+from graspy.utils import augment_diagonal, pass_to_ranks, import_graph, remove_loops, symmetrize
+
+from sklearn.utils.validation import check_is_fitted
+from sklearn.utils import check_array, check_X_y
+from sklearn.base import BaseEstimator
+
+
+def _check_common_inputs(
+    min_components, max_components, cluster_kws, embed_kws
+):
+    if not isinstance(min_components, int):
+        raise TypeError("min_components must be an int")
+    elif min_components < 1:
+        raise ValueError("min_components must be > 0")
+
+    if not isinstance(max_components, int):
+        raise TypeError("max_components must be an int")
+    elif max_components < 1:
+        raise ValueError("max_components must be > 0")
+    elif max_components < min_components:
+        raise ValueError("max_components must be >= min_components")
+
+    if embed_kws is not None:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    def __init__(
+        self,
+        selection_criteria_GMM="bic",
+        cluster_method="GMM",
+        root_inds=None,
+        n_init=1,
+        parent=None,
+        min_components=1,
+        max_components=10,
+        cluster_kws={},
+        # labels=None,  # not expecting label_init?
+        min_split=1,
+        reembed=False,
+    ):
+
+        _check_common_inputs(
+            min_components, max_components, cluster_kws, embed_kws=None
+        )
+
+        if cluster_method not in ["GMM", "KMeans"]:
+            msg = "clustering method must be one of {GMM, Kmeans}"
+            raise ValueError(msg)
+
+        self.parent = parent
+        self.n_init = n_init
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria_GMM = selection_criteria_GMM
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.root_inds = root_inds
+        self.reembed = reembed
+
+    def fit(self, X, adj=None):
+        self.fit_predict(X)
+        return self
+
+    def predict(self, X, adj=None):
```

Comment:
note that this does not use X, which is not what we want. X could be data that was not seen during fit